internal/app/machined/pkg/adapters/network/nftables_rule.go

```diff
@@ -0,0 +1,536 @@
+						Invert:         match.Invert,
+					},
+				)
+			}
+		}
+
+		// skip v6 rule if there are not IPs to match for and not inverted
+		if v6Set != nil || match.Invert {
+			if v6Set == nil && match.Invert {
+				// match any v6 IP
+				if rule6 == nil {
+					rule6 = []expr.Any{}
+				}
+			} else {
+				// match specific v6 IPs
+				result.Sets = append(result.Sets,
+					NfTablesSet{
+						Kind:      SetKindIPv6,
+						Addresses: v6Set,
+					})
+
+				rule6 = append(rule6,
+					// Store the destination IP address to register 1
+					&expr.Payload{
+						DestRegister: 1,
+						Base:         expr.PayloadBaseNetworkHeader,
+						Offset:       offV6,
+						Len:          16,
+					},
+					// Match from target set
+					&expr.Lookup{
+						SourceRegister: 1,
+						SetID:          uint32(len(result.Sets) - 1), // reference will be fixed up by the controller
+						Invert:         match.Invert,
+					},
+				)
+			}
+		}
+
+		return nil
+	}
+
+	if a.NfTablesRule.MatchSourceAddress != nil {
+		match := a.NfTablesRule.MatchSourceAddress
+
+		if err := addressMatchExpression(match, "source", 12, 8); err != nil {
+			return nil, err
+		}
+	}
+
+	if a.NfTablesRule.MatchDestinationAddress != nil {
+		match := a.NfTablesRule.MatchDestinationAddress
+
+		if err := addressMatchExpression(match, "destination", 16, 24); err != nil {
+			return nil, err
+		}
+	}
+
+	if a.NfTablesRule.MatchLayer4 != nil {
+		match := a.NfTablesRule.MatchLayer4
+
+		rulePre = append(rulePre,
+			// [ meta load l4proto => reg 1 ]
+			&expr.Meta{
+				Key:      expr.MetaKeyL4PROTO,
+				Register: 1,
+			},
+			// [ cmp eq reg 1 <protocol> ]
+			&expr.Cmp{
+				Op:       expr.CmpOpEq,
+				Register: 1,
+				Data:     []byte{byte(match.Protocol)},
+			},
+		)
+
+		portMatch := func(off uint32, ports []network.PortRange) {
+			result.Sets = append(result.Sets,
+				NfTablesSet{
+					Kind:  SetKindPort,
+					Ports: xslices.Map(ports, func(r network.PortRange) [2]uint16 { return [2]uint16{r.Lo, r.Hi} }),
+				},
+			)
+
+			rulePost = append(rulePost,
+				// [ payload load 2b @ transport header + <offset> => reg 1 ]
+				&expr.Payload{
+					DestRegister: 1,
+					Base:         expr.PayloadBaseTransportHeader,
+					Offset:       off,
+					Len:          2,
+				},
+				// [ lookup reg 1 set <set> ]
+				&expr.Lookup{
+					SourceRegister: 1,
+					SetID:          uint32(len(result.Sets) - 1), // reference will be fixed up by the controller
+				},
+			)
+		}
+
+		if match.MatchSourcePort != nil {
+			portMatch(0, match.MatchSourcePort.Ranges)
+		}
+
+		if match.MatchDestinationPort != nil {
+			portMatch(2, match.MatchDestinationPort.Ranges)
+		}
+	}
+
+	clampMSS := func(family nftables.TableFamily, mtu uint16) []expr.Any {
+		var mss uint16
+
+		switch family { //nolint:exhaustive
+		case nftables.TableFamilyIPv4:
+			mss = mtu - 40 // TCP + IPv4 overhead
+		case nftables.TableFamilyIPv6:
+			mss = mtu - 60 // TCP + IPv6 overhead
```

Comment:
would this change if wireguard is in use?

Reply:
so the MTU passed here via KubeSpan includes Wireguard overhead. In general this rule will clamp MSS to fit MTU